src/coreclr/jit/lsraxarch.cpp

```diff
@@ -2328,48 +2328,77 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree)
 
-                if (op2->isContained())
+                if (LIR::AsRange(blockSequence[curBBSeqNum]).TryGetUse(intrinsicTree, &use))
                 {
-                    // 132 form: op1 = (op1 * op3) + [op2]
+                    user = use.User();
+                }
+                resultOpNum = intrinsicTree->GetResultOpNumForFMA(user, op1, op2, op3);
 
-                    tgtPrefUse = BuildUse(op1);
+                // Intrinsics with CopyUpperBits semantics cannot have op1 be contained
+                assert(!copiesUpperBits || !op1->isContained());
 
-                    srcCount += 1;
-                    srcCount += BuildOperandUses(op2);
-                    srcCount += BuildDelayFreeUses(op3, op1);
+                unsigned containedOpNum = 0;
+
+                if (op1->isContained() || op1->IsRegOptional())
+                {
+                    containedOpNum = 1;
                 }
-                else if (op1->isContained())
+                else if (op2->isContained() || op2->IsRegOptional())
                 {
-                    // 231 form: op3 = (op2 * op3) + [op1]
-
-                    tgtPrefUse = BuildUse(op3);
-
-                    srcCount += BuildOperandUses(op1);
-                    srcCount += BuildDelayFreeUses(op2, op1);
-                    srcCount += 1;
+                    containedOpNum = 2;
                 }
                 else
                 {
-                    // 213 form: op1 = (op2 * op1) + [op3]
+                    assert(op3->isContained() || op3->IsRegOptional());
+                    containedOpNum = 3;
+                }
 
-                    tgtPrefUse = BuildUse(op1);
-                    srcCount += 1;
+                GenTree* emitOp1 = op1;
+                GenTree* emitOp2 = op2;
+                GenTree* emitOp3 = op3;
 
-                    if (copiesUpperBits)
+                // Intrinsics with CopyUpperBits semantics must have op1 as target
+                if (containedOpNum == 1 && !copiesUpperBits)
+                {
+                    if (resultOpNum != 3)
                     {
-                        srcCount += BuildDelayFreeUses(op2, op1);
+                        // op2 = ([op1] * op2) + op3
+                        std::swap(emitOp2, emitOp3);
                     }
-                    else
+
+                    // else: op3 = ([op1] * op2) + op3
+                    std::swap(emitOp1, emitOp3);
+                }
+                else if (containedOpNum == 3)
+                {
+                    if (resultOpNum == 2 && !copiesUpperBits)
                     {
-                        tgtPrefUse2 = BuildUse(op2);
-                        srcCount += 1;
+                        // op2 = (op1 * op2) + [op3]
+                        std::swap(emitOp1, emitOp2);
                     }
+                    // else: op1 = (op1 * op2) + [op3]
+                }
+                else
+                {
+                    assert(containedOpNum == 2);
```

Comment:
I think its possible for `containedOpNum` to be `0` and so we should check for this explicitly.